src/histogram.c

```diff
@@ -44,45 +267,119 @@ hist_sfunc(PG_FUNCTION_ARGS)
 	double		max = DatumGetFloat8(max_datum);
 	int			nbuckets = DatumGetInt32(nbuckets_datum);
 	int32		bucket = DatumGetInt32(DirectFunctionCall4(width_bucket_float8, val_datum, min_datum, max_datum, nbuckets_datum));
+	char	   *subagg_text;
 
+
+	/* Cannot be called directly because of internal-type argument */
 	if (!AggCheckCallContext(fcinfo, &aggcontext))
-	{
-		/* cannot be called directly because of internal-type argument */
 		elog(ERROR, "hist_sfunc called in non-aggregate context");
-	}
 
+	/* Cannot generate a histogram with incompatible bounds */
 	if (min > max)
-	{
-		/* cannot generate a histogram with incompatible bounds */
 		elog(ERROR, "lower bound cannot exceed upper bound");
-	}
+
+	/* Check if bucket_agg is set, otherwise use 'count' */
+	if (PG_NARGS() == 5)
+		subagg_text = "count";
+	else
+		subagg_text = TextDatumGetCString(PG_GETARG_CSTRING(5));
 
 	if (state == NULL)
 	{
-		/* Allocate memory to a new histogram state array */
-		Size		arrsize = sizeof(Datum) * (nbuckets + 2);
+		/* Length is nbuckets plus 1 on either side */
+		Size		arrsize = nbuckets + 2;
+		HistogramInternalState *istate =
+		(HistogramInternalState *) fcinfo->flinfo->fn_extra;
```

Comment:
I don't recall why this isn't part of or tagged on to the `HistogramState`?

Reply:
Also, if `HistogramInternalState` is in `fn_extra`, perhaps a better name is `HistogramInternalFunctionState` or something to indicate it is basically a function call cache. Not a biggie though.

Reply:
Done